gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java

```diff
@@ -95,6 +111,56 @@ public static Builder newBuilder() {
         .setUsingGDCH(false);
   }
 
+  /** Check that the User configured universe domain matches the Credentials' universe domain */
+  public boolean hasValidUniverseDomain(Credentials credentials) throws IOException {
+    if (usingGDCH()) {
+      // GDC-H has no universe domain, always return true.
+      return true;
+    } else if (credentials == null) {
+      // If passed with NoCredentialsProvider, match with GDU
+      return resolvedUniverseDomain().equals(Credentials.GOOGLE_DEFAULT_UNIVERSE);
+    }
+    return resolvedUniverseDomain().equals(credentials.getUniverseDomain());
+  }
+
+  public EndpointContext merge(EndpointContext input) {
```

Comment:
I'm not sure we need a default `EndpointContext` or this `merge` method. The main reason we need one for `ApiCallContext` is that, comparing to the default `ApiCallContext` created from `ClientContext`, there could be different value for some settings(e.g. RetrySettings) for different RPCs. But I don't think we would have different    `EndpointContext` for different RPCs.